compiler/rustc_mir/src/transform/instcombine.rs

```diff
@@ -87,6 +93,49 @@ impl OptimizationFinder<'b, 'tcx> {
         OptimizationFinder { body, tcx, optimizations: OptimizationList::default() }
     }
 
+    fn find_deref_of_address(&mut self, rvalue: &Rvalue<'tcx>, location: Location) -> Option<()> {
+        // look for the sequence
+        // _2 = &_1;
+        // ...
+        // _5 = (*_2)
+        //
+        // which we can replace the last statement with `_5 = _1` to avoid the load of _2
+        if let Rvalue::Use(op) = rvalue {
+            let local_being_derefed = match op.place()?.as_ref() {
+                PlaceRef { local, projection: [ProjectionElem::Deref] } => Some(local),
+                _ => None,
+            }?;
+
+            let stmt_index = location.statement_index;
+            // look behind for statement that assigns the local from a address of operator
+            let statements_to_look_in =
+                self.body.basic_blocks()[location.block].statements[0..stmt_index].iter().rev();
+            for stmt in statements_to_look_in {
+                match &stmt.kind {
+                    rustc_middle::mir::StatementKind::Assign(box (l, r))
+                        if l.local == local_being_derefed =>
+                    {
+                        match r {
+                            Rvalue::Ref(
+                                _,
+                                BorrowKind::Shared | BorrowKind::Shallow,
+                                place_taken_address_of,
+                            ) => {
+                                self.optimizations
+                                    .unneeded_deref
+                                    .insert(location, *place_taken_address_of);
+                                return Some(());
+                            }
+                            _ => {}
+                        }
+                    }
+                    _ => {}
```

Comment:
What happens in a code sequence like this?

```rust
_1 = &_2;
_1 = _3;
_4 = *_1;
```

If I'm reading the code correctly, this would replace `_4 = *_1` with `_4 = _2`, which is incorrect.

Reply:
Is it ever possible to have that sequence? I thought SSA form guaranteed only a single assignment to the same variable.

Reply:
MIR is not in SSA form

Reply:
In that case, yeah, that sequence will get miscompiled. I think it can be fairly easy fixed by bailing if an assignment to _1 is seen that is not of the expected &_1 form. I'll push that up.